### PR TITLE
Fixed incorrect discard suggestion in var pattern

### PR DESCRIPTION
**Pull request by DoctorKrolic**
Fixes: https://github.com/dotnet/roslyn/issues/45768

### CODE REVIEW

File: src/Analyzers/CSharp/Tests/RemoveUnusedParametersAndValues/RemoveUnusedValueAssignmentTests.cs [hunk 1]
```diff
@@ -8518,7 +8518,7 @@ void M(C c)
     public object P { get; }
     void M(C c)
     {
-        var x = c is { P : _ };
+        var x = c is { P : var _ };
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Analyzers/CSharp/Tests/RemoveUnusedParametersAndValues/RemoveUnusedValueAssignmentTests.cs:8521]: I would really prefer to keep the original behavior here.

File: src/Analyzers/VisualBasic/CodeFixes/RemoveUnusedParametersAndValues/VisualBasicRemoveUnusedValuesCodeFixProvider.vb [hunk 2]
```diff
@@ -63,6 +63,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.RemoveUnusedParametersAndValues
             Throw New NotImplementedException()
         End Function
 
+        Protected Overrides Function GetReplacementNodeForVarPattern(originalVarPattern As SyntaxNode, newNameNode As SyntaxNode) As SyntaxNode
+            ' VB does not have var patterns
+            Throw New NotImplementedException()
```
Comment (sharwell) [inline src/Analyzers/VisualBasic/CodeFixes/RemoveUnusedParametersAndValues/VisualBasicRemoveUnusedValuesCodeFixProvider.vb:68]:
```suggestion
            Throw ExceptionUtilities.Unreachable
```

File: src/Analyzers/CSharp/Tests/RemoveUnusedParametersAndValues/RemoveUnusedValueAssignmentTests.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Analyzers/CSharp/Tests/RemoveUnusedParametersAndValues/RemoveUnusedValueAssignmentTests.cs:8521]: It should be easy to account for this in `GetReplacementNodeForVarPattern` by just checking the parent kind.

File: src/Analyzers/VisualBasic/CodeFixes/RemoveUnusedParametersAndValues/VisualBasicRemoveUnusedValuesCodeFixProvider.vb (hunk 2 above)
Comment (DoctorKrolic) [reply src/Analyzers/VisualBasic/CodeFixes/RemoveUnusedParametersAndValues/VisualBasicRemoveUnusedValuesCodeFixProvider.vb:68]: Both of them, right?)

**Review by sharwell — CHANGES_REQUESTED** [review]
Requesting two changes:

1. Fix the trivia when `var _` is used
2. Preserve the use of `_` in cases where `var _` is unnecessary

File: src/Analyzers/VisualBasic/CodeFixes/RemoveUnusedParametersAndValues/VisualBasicRemoveUnusedValuesCodeFixProvider.vb (hunk 2 above)
Comment (sharwell) [reply src/Analyzers/VisualBasic/CodeFixes/RemoveUnusedParametersAndValues/VisualBasicRemoveUnusedValuesCodeFixProvider.vb:68]: You can do both, or restrict it to just the method you added.

File: src/Analyzers/CSharp/CodeFixes/RemoveUnusedParametersAndValues/CSharpRemoveUnusedValuesCodeFixProvider.cs [hunk 3]
```diff
@@ -164,5 +166,32 @@ protected override void InsertAtStartOfSwitchCaseBlockForDeclarationInCaseLabelO
                 return SyntaxFactory.BinaryExpression(mappedBinaryExpressionKind, leftOfAssignment, rightOfAssignment);
             }
         }
+
+        protected override SyntaxNode GetReplacementNodeForVarPattern(SyntaxNode originalVarPattern, SyntaxNode newNameNode)
+        {
+            if (originalVarPattern is not VarPatternSyntax pattern)
+            {
+                Debug.Fail($"The given {nameof(originalVarPattern)} node is not a var pattern, but {originalVarPattern.Kind()}");
```
Comment (davidwengier) [inline src/Analyzers/CSharp/CodeFixes/RemoveUnusedParametersAndValues/CSharpRemoveUnusedValuesCodeFixProvider.cs:174]: Seems like this, and the one at the end of the method, should be `throw ExceptionUtilities.Unreachable`. A debug fail is most likely only going to show up when running tests, and I assume you haven't written any tests that hit this line, nor would any existing tests.

Comment (DoctorKrolic) [reply src/Analyzers/CSharp/CodeFixes/RemoveUnusedParametersAndValues/CSharpRemoveUnusedValuesCodeFixProvider.cs:174]: > you haven't written any tests that hit this line, nor would any existing tests.

I don't think it is possible. Unreachable exception indeed seems to better here

**Review by davidwengier — APPROVED** [review]

**Comment by davidwengier** [comment]
Thanks

**Comment by sharwell** [comment]
The CI build after merge is showing an error like this:

> Still unable to load MEF component DLL: Could not load file or assembly 'Microsoft.VisualStudio.LanguageServer.Client, Version=17.3.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.

/cc @dibarbet

**Pull request merged**